Ansible Deployment for Windows

Mukeshr-Ram opened the pull request:
Install splunk-otel-collector using Ansible for Windows following MSI installer approach.

Downloading and installing Splunk-otel collector and FluentD for Windows.
Setting up appropriate Registry Values for Windows.
Updated README.md.

flands commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,16 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
+## Windows
+Currently, the following Windows versions are supported:
+Ansible requires PowerShell 3.0 or newer and atleast .NET4.0 to be installed on Windows host.
+A WinRM listner should be created and activeted. 
```
Not a Windows/Ansible expert -- not sure what this is or how to ensure available

Mukeshr-Ram replied:
For detailed setup of Ansible in windows I have added Ansible docs reference link.

flands commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,16 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
 
+## Windows
+Currently, the following Windows versions are supported:
+Ansible requires PowerShell 3.0 or newer and atleast .NET4.0 to be installed on Windows host.
```
```suggestion
- Ansible requires PowerShell 3.0 or newer and at least .NET 4.0 to be installed on Windows host.
```

flands commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,16 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
+## Windows
+Currently, the following Windows versions are supported:
+Ansible requires PowerShell 3.0 or newer and atleast .NET4.0 to be installed on Windows host.
+A WinRM listner should be created and activeted. 
```
```suggestion
- A WinRM listener should be created and activated. 
```

flands commented on deployments/ansible/roles/collector/README.md:
````diff
@@ -22,6 +32,8 @@ how to use the role in a playbook with minimal required configuration:
 ```yaml
 - name: Install Splunk OpenTelemetry Connector
   hosts: all
+  #For Windows "become: yes" will raise error.
+  #"The Powershell family is incompatible with the sudo become plugin" Remove "become: yes" tag to run on Windows
````
```suggestion
  # For Windows "become: yes" will raise error.
  # "The Powershell family is incompatible with the sudo become plugin" Remove "become: yes" tag to run on Windows
```

flands commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,16 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
 collect metrics, traces and logs from Linux machines and send data to [Splunk 
 Observability Cloud](https://www.splunk.com/en_us/observability.html). 
 
+## Windows
```
It is confusing that a Windows header exists, but not a Linux -- made me initially think this page was just for Windows. If we add for Windows we should add for Linux as well and section header generic.

Mukeshr-Ram replied:
Yes, I will add Linux as well.

flands commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -63,20 +75,25 @@ how to use the role in a playbook with minimal required configuration:
   on the node. The `SPLUNK_BUNDLE_DIR` environment variable will be set to
   this value for the collector service.  (**default:**
-  `/usr/lib/splunk-otel-collector/agent-bundle`)
+  `/usr/lib/splunk-otel-collector/agent-bundle` For Windows `${collector_path_win}\agent-bundle`)
```
This Linux versus Windows thing is hard to read -- perhaps all these options can be turned into a table

Parameter (default) | OS | Description

Where OS is Linux|Windows|All

flands replied:
Another alternative is to have different sections per OS -- this would result in duplicated information, but perhaps fine as it make it more clear

Mukeshr-Ram replied:
Sure will update in that format.

flands commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -34,3 +34,38 @@ splunk_fluentd_config_source: ""
+
+package_stage: release
+# Splunk-otel-collector Windows MSI version
+win_otel_version: 0.31.0
```
```suggestion
win_otel_version: latest
```

Mukeshr-Ram replied:
latest tag is not working in this case.

flands commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -34,3 +34,38 @@ splunk_fluentd_config_source: ""
+# Splunk-otel-collector Windows MSI version
+win_otel_version: 0.31.0
+# FluentD Windows Version
+win_td_agent_version: 4.1.1
```
Does latest work here?

Mukeshr-Ram replied:
No.

flands commented on deployments/ansible/roles/collector/handlers/main.yml:
```diff
@@ -20,3 +20,15 @@
+    state: restarted
+  listen: "restart windows splunk-otel-collector"
+
+- name: Restart Splunk td-agent for windows
```
```suggestion
- name: Restart td-agent for windows
```

flands commented on deployments/ansible/roles/collector/tasks/win_fluent_install.yml:
```diff
@@ -0,0 +1,33 @@
+    state: absent
+  when: service_info.exists
+
+- name: install Fluentd for windows
```
```suggestion
- name: Install Fluentd for windows
```

flands reviewed (commented):
Took a quick pass -- before merge also need to update top-level README as well as CHANGELOG

jcheng-splunk commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -105,8 +126,50 @@ how to use the role in a playbook with minimal required configuration:
+## Windows Specific Role Variables
+
+- `collector_path_win`: Default path of Splunk-otel-collector in windows 
+  (**default:** `C:\\Program Files\Splunk\OpenTelemetry Collector\`)
```
The MSI installation directory is not currently configurable and installs to `%PROGRAMFILES%\Splunk\OpenTelemetry Collector`.

jcheng-splunk commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -105,8 +126,50 @@ how to use the role in a playbook with minimal required configuration:
+- `collector_path_win`: Default path of Splunk-otel-collector in windows 
+  (**default:** `C:\\Program Files\Splunk\OpenTelemetry Collector\`)
+
+- `splunk_bundle_dir_win` & : The path to the [Smart Agent bundle directory](
```
Is the `&` needed here?

jcheng-splunk commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -105,8 +126,50 @@ how to use the role in a playbook with minimal required configuration:
+  variable will be set to this value for the collector service.
+  (**default:** `${splunk_bundle_dir_win}\run\collectd`)
+
 - `splunk_fluentd_config_source`: Source path to a fluentd config file on your 
```
Should this be moved out of the windows section?

jcheng-splunk commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -105,8 +126,50 @@ how to use the role in a playbook with minimal required configuration:
   that nothing will be copied and existing `splunk_fluentd_config` will be used)
+
+- `collector_config_source_win`: Source path to the collector config YAML file. This file will 
+  be copied to the $collector_config_dest path on the node. See the source attribute of the file 
```
The `See the source attribute of the file resource for supported value types.` statement is not applicable to ansible and should be removed.

jcheng-splunk commented on deployments/ansible/roles/collector/tasks/win_install.yml:
```diff
@@ -0,0 +1,34 @@
+    win_fluentd_msi: td-agent-{{win_td_agent_version}}-x64.msi
+
+# creating Temp dir for downloading MSI packages
+- name: Create directory structure
```
May not be needed if using https://docs.ansible.com/ansible/latest/collections/ansible/windows/win_tempfile_module.html.

Mukeshr-Ram commented:
I have committed suggested changes in the latest commit 

- Fetch latest version of splunk-otel-collector.
- Push custom config file from source to dest for fluentd & splunk-otel-collector.
- Updated ENV VARS.
- Updated Readme file.

Please Review and suggest any additional changes, if required.

jcheng-splunk commented on deployments/ansible/roles/collector/tasks/win_fluent_install.yml:
```diff
@@ -0,0 +1,47 @@
+
+- name: Set fluentd package sources for Windows
+  ansible.builtin.set_fact:
+    td_agent_base_url: http://packages.treasuredata.com.s3.amazonaws.com
```
```suggestion
    td_agent_base_url: https://packages.treasuredata.com
```

rmfitzpatrick commented on deployments/ansible/README.md:
````diff
@@ -24,6 +42,8 @@ Note that this role requires root access.
 ```yaml
 - name: Install Splunk OpenTelemetry Connector
   hosts: all
+  # For Windows "become: yes" will raise error.
````
What error are you seeing?  become should be supported though we may need to add a platform-dependent `become_user`?: https://docs.ansible.com/ansible/latest/user_guide/become.html#id15

Mukeshr-Ram replied:
Yes, we can add become_user for windows

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -34,3 +34,42 @@ splunk_fluentd_config_source: ""
+
+# default vars For Windows family
+
+splunk_api_url: https://api.{{splunk_realm}}.signalfx.com
+
+splunk_ingest_url:  https://ingest.{{splunk_realm}}.signalfx.com
+
+splunk_trace_url: "{{splunk_ingest_url}}/v2/trace"
+
+splunk_hec_url: "{{splunk_ingest_url}}/v1/log"
+
+splunk_hec_token: "{{splunk_access_token}}"
+
```
These variables are not be Windows specific by definition. Can we remove them to avoid confusion? We can add them later for both linux and windows if needed

Mukeshr-Ram replied:
For Linux these vars are getting fetched from templates and getting copied into splunk-otel-collector.conf. Let me try creating a separate file for Windows vars, if it works I will commit those changes.

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -34,3 +34,42 @@ splunk_fluentd_config_source: ""
+
+splunk_hec_token: "{{splunk_access_token}}"
+
+splunk_bundle_dir_win: "{{ ansible_env.ProgramFiles }}/Splunk/OpenTelemetry Collector/agent-bundle"
+
+splunk_collectd_dir_win: "{{splunk_bundle_dir_win}}/run/collectd"
+
+collector_config_source_win:  "{{ ansible_env.ProgramFiles }}/Splunk/OpenTelemetry Collector/agent_config.yaml"
+
+collector_config_dest_win: "{{ ansible_env.ProgramData }}/Splunk/OpenTelemetry Collector/agent_config.yaml"
+
+win_fluentd_config_source: "%{{ ansible_env.ProgramFiles }}/Splunk/OpenTelemetry Collector/fluentd/td-agent.conf"
+
+win_fluentd_config_dest: "{{ ansible_env.SystemDrive }}/opt/td-agent/etc/td-agent/td-agent.conf"
+
+package_stage: release
+
+windows_base_url: https://dl.signalfx.com
+
+win_otel_msi: splunk-otel-collector-{{win_otel_version}}-amd64.msi
+
+# Splunk-otel-collector Windows MSI version
+win_otel_version: latest
+
+# FluentD Windows Version
+win_td_agent_version: 4.1.1
+
```
What is the reason to have completely different set of config option for Windows? Can we reuse existing set of options and make sure the defaults are set correctly based on the platform?

Let's add extra options only if it's really necessary specifically for Windows.

We should start with default behaviour and add more configuration options (ideally independent from the platform) only when it's really needed for users, otherwise, if we provide a lot of unnecessary options, the configuration interface becomes pretty convoluted.

Mukeshr-Ram replied:
For Windows the config paths, versions and installation approach are different from linux so, i have used a different set of config options for Windows. I am not sure whether we can able to reuse the existing set of options because the existing defaults are linux dependent. In this case we can create two different files in defaults based on the OS platform one for linux the other for windows and need to specify the vars file using 'include_vars:' option in the playbook based on OS platform before running the ansible role.

rmfitzpatrick commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -105,8 +126,40 @@ how to use the role in a playbook with minimal required configuration:
 - `splunk_fluentd_config`: Path to the fluentd config file on the remote host.
   (**default:** `/etc/otel/collector/fluentd/fluent.conf`)
 
-- `splunk_fluentd_config_source`: Source path to a fluentd config file on your 
```
This is still an active variable? Why is the description being removed?

Mukeshr-Ram replied:
Added back.

rmfitzpatrick commented on deployments/ansible/roles/collector/vars/main.yml:
```diff
@@ -33,4 +33,4 @@ splunk_fluentd_config: /etc/otel/collector/fluentd/fluent.conf
 
 agent_bundle_dir: /usr/lib/splunk-otel-collector/agent-bundle
-splunk_collectd_dir: /usr/lib/splunk-otel-collector/agent-bundle/run/collectd
+splunk_collectd_dir: /usr/lib/splunk-otel-collector/agent-bundle/run/collectd
```
nit
```suggestion
splunk_collectd_dir: /usr/lib/splunk-otel-collector/agent-bundle/run/collectd

```

dmitryax commented:
Please add windows to the CI https://github.com/signalfx/splunk-otel-collector/blob/main/.github/workflows/ansible.yml#L55 and rebase on the latest `main`

dmitryax commented on deployments/ansible/README.md:
```diff
@@ -24,7 +42,7 @@ Note that this role requires root access.
 - name: Install Splunk OpenTelemetry Connector
   hosts: all
-  become: yes
+  become_user: root
```
Why is this changed?

Mukeshr-Ram replied:
For Windows Its giving a error 'The Powershell family is incompatible with the sudo become plugin' so, changed to 'become_user: root'.

dmitryax commented on deployments/ansible/README.md:
```diff
@@ -8,6 +8,24 @@ The role installs Splunk OpenTelemetry Connector configured to
 collect metrics, traces and logs from Linux machines and send data to [Splunk 
 Observability Cloud](https://www.splunk.com/en_us/observability.html). 
 
+## Linux
+Currently, the following Linux distributions and versions are supported:
+
+- Amazon Linux
+- RedHat
+- Debian: buster, stretch
+- Ubuntu: 16.04, 18.04, 20.04
+
+## Windows
+Currently, the following Windows versions are supported:
+Ansible requires PowerShell 3.0 or newer and atleast .NET4.0 to be installed on Windows host.
+A WinRM listner should be created and activeted. 
+For setting up Windows Host Refer:[Ansible Docs](https://docs.ansible.com/ansible/latest/user_guide/windows_setup.html).
+
+- Windows Server 2012 64-bit
+- Windows Server 2016 64-bit
+- Windows Server 2019 64-bit
+
```
No need to duplicate this info here and in the role's README. Let's remove it from here

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,24 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
+Currently, the following Windows versions are supported:
+- Ansible requires PowerShell 3.0 or newer and at least .NET 4.0 to be installed on Windows host.
+A WinRM listner should be created and activated. 
+For setting up Windows Host Refer : [Ansible Docs](https://docs.ansible.com/ansible/latest/user_guide/windows_setup.html)
```
```suggestion
For setting up Windows Host refer [Ansible Docs](https://docs.ansible.com/ansible/latest/user_guide/windows_setup.html)
```

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,24 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
+## Windows
+Currently, the following Windows versions are supported:
+- Ansible requires PowerShell 3.0 or newer and at least .NET 4.0 to be installed on Windows host.
+A WinRM listner should be created and activated. 
```
```suggestion
A WinRM listener should be created and activated. 
```

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -4,6 +4,24 @@ Ansible role that installs Splunk OpenTelemetry Connector configured to
+A WinRM listner should be created and activated. 
+For setting up Windows Host Refer : [Ansible Docs](https://docs.ansible.com/ansible/latest/user_guide/windows_setup.html)
+
+- Windows Server 2012 64-bit
+- Windows Server 2016 64-bit
+- Windows Server 2019 64-bit
```
Can we move this list higher right under "Currently, the following Windows versions are supported:" and separate it from the list of Windows requirements

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -22,7 +40,7 @@ how to use the role in a playbook with minimal required configuration:
 - name: Install Splunk OpenTelemetry Connector
   hosts: all
-  become: yes
+  become_user: root
```
Same comment here. Why changed? AFAIK `become_user: root` is default anyway, but it doesn't work if there is no `become: yes`.

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -56,27 +74,35 @@ how to use the role in a playbook with minimal required configuration:
 - `splunk_otel_collector_config`: Splunk OTel Collector config YAML file. Can be set to 
   `/etc/otel/collector/gateway_config.yaml` to install the collector in gateway
   mode. (**default:** `/etc/otel/collector/agent_config.yaml`)
+  For Windows this is the destination path of the collector config file on the node. 
+  The SPLUNK_CONFIG environment variable will be set with this value for the collector service.
+  (**default:** `%ProgramData%\Splunk\OpenTelemetry Collector\agent_config.yaml`)
```
Why the definition is different for Windows? It should have the same meaning. So we can just change the default value to the following:
**default:** `/etc/otel/collector/agent_config.yaml` on Linux, `%ProgramData%\Splunk\OpenTelemetry Collector\agent_config.yaml` on Windows

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -1,16 +1,16 @@
+# Set this variable to PATH/gateway_config.yaml in order to
 # deploy splunk-otel-collector in gateway mode
-splunk_otel_collector_config: /etc/otel/collector/agent_config.yaml
+splunk_otel_collector_config: ""
```
Lets add comments here with default values for both platforms

Mukeshr-Ram replied:
Okay.

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -29,8 +29,13 @@ install_fluentd: true
-splunk_fluentd_config: /etc/otel/collector/fluentd/fluent.conf
+splunk_fluentd_config: ""
 splunk_fluentd_config_source: ""
 
-agent_bundle_dir: /usr/lib/splunk-otel-collector/agent-bundle
-splunk_collectd_dir: /usr/lib/splunk-otel-collector/agent-bundle/run/collectd
+agent_bundle_dir: ""
+splunk_collectd_dir: ""
```
Same ask here, let's add comments with default values

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -56,27 +74,35 @@ how to use the role in a playbook with minimal required configuration:
   that nothing will be copied and existing `splunk_otel_collector_config` will be used)
-
+  For Windows this file will be copied to the $splunk_otel_collector_config path on the node.
+  (**default:** `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent_config.yaml`)
```
Same Q as above, another definition for Windows is confusing. The parameter should have the same meaning and behaviour across platforms

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -56,27 +74,35 @@ how to use the role in a playbook with minimal required configuration:
   on the node. The `SPLUNK_BUNDLE_DIR` environment variable will be set to
-  this value for the collector service.  (**default:**
-  `/usr/lib/splunk-otel-collector/agent-bundle`)
+  this value for the collector service. For Linux (**default:**
+  `/usr/lib/splunk-otel-collector/agent-bundle`). For Windows (**default:** 
+  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle`)
```
```suggestion
  this value for the collector service. (**default:**
  `/usr/lib/splunk-otel-collector/agent-bundle` on Linux,  
  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle` on Windows)
```

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -103,10 +129,13 @@ how to use the role in a playbook with minimal required configuration:
 
 - `splunk_fluentd_config`: Path to the fluentd config file on the remote host.
-  (**default:** `/etc/otel/collector/fluentd/fluent.conf`)
+  For Linux(**default:** `/etc/otel/collector/fluentd/fluent.conf`). 
+  For Windows (**default:** `%SYSTEMDRIVE%\opt\td-agent\etc\td-agent\td-agent.conf`)
```
```suggestion
  (**default:** `/etc/otel/collector/fluentd/fluent.conf` on Linux, 
  `%SYSTEMDRIVE%\opt\td-agent\etc\td-agent\td-agent.conf` on Windows)
```

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -56,27 +74,35 @@ how to use the role in a playbook with minimal required configuration:
   an existing directory on the node. The `SPLUNK_COLLECTD_DIR` environment
-  variable will be set to this value for the collector service.  (**default:**
-  `/usr/lib/splunk-otel-collector/agent-bundle`)
+  variable will be set to this value for the collector service. For Linux
+  (**default:** `/usr/lib/splunk-otel-collector/agent-bundle`). For Windows(**default:** 
+  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd`)
```
```suggestion
  variable will be set to this value for the collector service. (**default:** 
  `/usr/lib/splunk-otel-collector/agent-bundle` on Linux, 
  `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd` on Windows)
```

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -103,10 +129,13 @@ how to use the role in a playbook with minimal required configuration:
-  e.g. `./custom_fluentd_config.conf`. (**default:** `""` meaning 
+  e.g. `./custom_fluentd_config.conf`. For Linux (**default:** `""` meaning 
   that nothing will be copied and existing `splunk_fluentd_config` will be used)
+  For Windows this file will be copied to the $splunk_fluentd_config path on the node.
+  (**default:** `%ProgramFiles%\Splunk\OpenTelemetry Collector\fluentd\td-agent.conf`)
```
Same comment

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -29,8 +29,13 @@ install_fluentd: true
+agent_bundle_dir: ""
+splunk_collectd_dir: ""
+
+package_stage: release
```
It's not configurable on Linux, let's do not add it for Windows

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -29,8 +29,13 @@ install_fluentd: true
+
+package_stage: release
+
+msi_name: splunk-otel-collector-{{splunk_otel_collector_version}}-amd64.msi
+win_base_url: https://dl.signalfx.com
```
Same for these two. No need to make it configurable at this point

Mukeshr-Ram replied:
msi_name should be configurable to fetch latest version.

dmitryax replied:
Why? Can't the version be configurable with `splunk_otel_collector_version`?

Mukeshr-Ram replied:
Hello @dmitryax, When we set msi_name using set_fact it's only fetching the version as "latest" from defaults/main.yml. The latest version we are fetching from dl.signalfx.com is not overriding the default value "latest". We can give the name of msi directly in the "URL" but lint test is failing due to error line too long.

dmitryax replied:
> We can give the name of msi directly in the "URL" but lint test is failing due to error line too long.

You can just break down the line as done here for example https://github.com/signalfx/splunk-otel-collector/blob/5f2b860291a4872e868961cab809d629e724bd69/deployments/ansible/roles/collector/tasks/yum_install_fluentd.yml#L12-L13

dmitryax commented on deployments/ansible/roles/collector/tasks/main.yml:
```diff
@@ -1,13 +1,29 @@
 ---
+- name: Define acceptable distros For splunk-otel-collector
+  set_fact:
+    debian_distro: ['Debian', 'Ubuntu']
+    rhel_distro: ['CentOS', 'RedHat', 'Red Hat Enterprise Linux', 'Amazon']
```
I don't see these facts being used. Doesn't seem like this task is needed at all

Mukeshr-Ram replied:
Ok I will remove it.

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+# Set Vars for Linux deployments.
+- name: set vars for linux
+  set_fact:
+    # Set this variable to /etc/otel/collector/gateway_config.yaml in order to
+    # deploy splunk-otel-collector in gateway mode
```
We don't need this comment here

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+      {%- else -%}
+        /etc/otel/collector/agent_config.yaml
+      {%- endif -%}
+    # Set splunk_otel_collector_config_source to copy a custom config from
+    # a local path to splunk_otel_collector_config on remote hosts
```
Same. These comments should be in defaults/main.yml

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+---
+# Set Vars for Linux deployments.
+- name: set vars for linux
```
```suggestion
- name: Set default vars for Linux
```

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+      {%- endif -%}
+  when: not ansible_os_family in windows_distro
+
+# Set Vars for Windows deployments.
+- name: set vars for windows
```
```suggestion
# Set Vars for Windows deployments.
- name: Set default vars for Windows
```

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+      {%- else -%}
+        {{ansible_env.ProgramFiles}}/Splunk/OpenTelemetry Collector/agent-bundle/run/collectd
+      {%- endif -%}
+    # In Windows to deploy splunk-otel-collector in gateway mode replace agent_config.yaml with gateway_config.yaml in
+    # 'splunk_otel_collector_config' 'splunk_otel_collector_config_source' this will copy gateway_config from source.
```
Confusing comment, please remove

dmitryax commented on deployments/ansible/roles/collector/tasks/otel_win_install.yml:
```diff
@@ -0,0 +1,53 @@
+  ansible.windows.win_copy:
+    src: "{{splunk_otel_collector_config_source}}"
+    dest: "{{splunk_otel_collector_config}}"
+    remote_src: yes
```
It shouldn't be `remote_src`. `splunk_otel_collector_config_source` is meant to upload a local custom config to a remote host

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+      {%- if splunk_otel_collector_config_source -%}
+        {{ splunk_otel_collector_config_source }}
+      {%- else -%}
+         {{ansible_env.ProgramFiles}}/Splunk/OpenTelemetry Collector/agent_config.yaml
```
`splunk_otel_collector_config_source` should be empty by default. It's needed only if user wants to upload a local custom config to a remote host. this variable shouldn't be modified here for any platform

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+      {%- else -%}
+        /etc/otel/collector/fluentd/fluent.conf
+      {%- endif -%}
+    splunk_fluentd_config_source: "{{ splunk_fluentd_config_source }}"
```
No reason to redefine it here

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,85 @@
+      {%- if splunk_fluentd_config -%}
+        {{ splunk_fluentd_config }}
+      {%- else -%}
+        {{ansible_env.SystemDrive}}/opt/td-agent/etc/td-agent/td-agent.conf
```
Same comment as for `splunk_otel_collector_config_source`

Mukeshr-Ram commented:
Hello @dmitryax 
Updated the PR with suggested changes.
Can you please review it.
Also, guide me if any other changes are needed on this PR.

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -55,28 +57,32 @@ how to use the role in a playbook with minimal required configuration:
-  config, e.g. `./custom_collector_config.yaml`. (**default:** `""` meaning 
-  that nothing will be copied and existing `splunk_otel_collector_config` will be used)
-
+  config, e.g. `./custom_collector_config.yaml`. (**default:** `""`) 
+  meaning that nothing will be copied and existing `splunk_otel_collector_config` will be used)
+  
```
Please discard this change, looks unrelated

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -110,10 +116,11 @@ how to use the role in a playbook with minimal required configuration:
   remote hosts. Can be used to submit a custom fluentd config,
-  e.g. `./custom_fluentd_config.conf`. (**default:** `""` meaning 
-  that nothing will be copied and existing `splunk_fluentd_config` will be used)
+  e.g. `./custom_fluentd_config.conf`. (**default:** `""`)
+  meaning that nothing will be copied and existing `splunk_fluentd_config` will be used)
```
Same comment, parentheses mismatch

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -1,16 +1,18 @@
 ---
 # defaults file for splunk-otel-collector role vars
-
+# Note: To modify vars based upon ansible_os_family  change the values of vars in tasks/vars.yml
```
The comment looks confusing. Why user would have to change values in tasks/vars.yml instead of using the provided set of config options?

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -1,16 +1,18 @@
-# deploy splunk-otel-collector in gateway mode
-splunk_otel_collector_config: /etc/otel/collector/agent_config.yaml
+# Set this variable to /etc/otel/collector/gateway_config.yaml on linux,
+# "{{ansible_env.ProgramFiles}}/Splunk/OpenTelemetry Collector/gateway_config.yaml" on Windows
```
Shouldn't it be `%ProgramFiles%/Splunk/OpenTelemetry Collector/gateway_config.yaml`? I don't think that jinja template evaluation will be applied on user input

Mukeshr-Ram replied:
If we use ` %ProgramFiles%/Splunk/OpenTelemetry Collector/gateway_config.yaml ` Windows reading this value as a literal, we need this value to be assigned as ENV var for SPLUNK_CONFIG. So, I have used 
` {{ansible_env.ProgramFiles}}/Splunk/OpenTelemetry Collector/gateway_config.yaml` to resolve that issue.

dmitryax replied:
But if user uses `{{ansible_env.ProgramFiles}}/Splunk/OpenTelemetry Collector/gateway_config.yaml` value as is in `splunk_otel_collector_config` ansible will likely use the value as is without evaluating `ansible_env.ProgramFiles`, so it won't work. We probably should specify something like `Program Files\Splunk\OpenTelemetry Collector\gateway_config.yaml`

dmitryax commented on deployments/ansible/roles/collector/tasks/win_install.yml:
```diff
@@ -0,0 +1,18 @@
+# Install splunk-otel-collector on Windows
+- name: Install Splunk OpenTelemetry Collector with msi package manager for Windows
+  ansible.builtin.import_tasks: otel_win_install.yml
+  when: ansible_os_family == "Windows"
```
`when: ansible_os_family == "Windows"` conditions are not needed in this file since it's imported with the same condition

dmitryax commented on deployments/ansible/roles/collector/tasks/win_fluentd_install.yml:
```diff
@@ -0,0 +1,27 @@
+    path: "{{fluentd_msi.dest}}"
+    state: present
+
+# Push Custom conguration file for fluentd
```
Please remove all the comments above the tasks. Task name supposed to be descriptive enough. No need to duplicate it

dmitryax commented on deployments/ansible/roles/collector/tasks/win_install.yml:
```diff
@@ -0,0 +1,18 @@
+  when: ansible_os_family == "Windows"
+
+# Install fluentd on Windows
+- name: Install Splunk OpenTelemetry Collector with yum package manager
```
```suggestion
- name: Install Fluentd with msi package manager
```

Mukeshr-Ram commented:
Hello @dmitryax
Updated the PR with suggested changes.
Can you please review it.

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -55,28 +57,32 @@ how to use the role in a playbook with minimal required configuration:
-  config, e.g. `./custom_collector_config.yaml`. (**default:** `""` meaning 
-  that nothing will be copied and existing `splunk_otel_collector_config` will be used)
-
+  config, e.g. `./custom_collector_config.yaml`. (**default:** `""` 
+  meaning that nothing will be copied and existing `splunk_otel_collector_config` will be used)
+  
```
I think I commented on this already. Please discard this change

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -55,28 +57,32 @@ how to use the role in a playbook with minimal required configuration:
+  variable will be set to this value for the collector service.
+  (**default:** `/usr/lib/splunk-otel-collector/agent-bundle` on Linux, 
+  **default:** `%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd` on Windows)
+
```
nit: extra space

dmitryax commented on deployments/ansible/roles/collector/README.md:
```diff
@@ -110,10 +116,11 @@ how to use the role in a playbook with minimal required configuration:
   remote hosts. Can be used to submit a custom fluentd config,
-  e.g. `./custom_fluentd_config.conf`. (**default:** `""` meaning 
-  that nothing will be copied and existing `splunk_fluentd_config` will be used)
+  e.g. `./custom_fluentd_config.conf`. (**default:** `""` 
+  meaning that nothing will be copied and existing `splunk_fluentd_config` will be used)
```
Please discard this change. Try to keep you diffs always related to the change you are working on

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -29,11 +31,18 @@ install_fluentd: true
 td_agent_version: ""
 
-splunk_fluentd_config: /etc/otel/collector/fluentd/fluent.conf
+# Default path /etc/otel/collector/fluentd/fluent.conf on Linux.
+# Default path %ProgramFiles%/Splunk/OpenTelemetry Collector/fluentd/td-agent.conf" on Windows.
```
```suggestion
# Default path on Linux: "/etc/otel/collector/fluentd/fluent.conf".
# Default path on Windows: "%ProgramFiles%\Splunk\OpenTelemetry Collector\fluentd\td-agent.conf".
```

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -29,11 +31,18 @@ install_fluentd: true
 
-agent_bundle_dir: /usr/lib/splunk-otel-collector/agent-bundle
-splunk_collectd_dir: /usr/lib/splunk-otel-collector/agent-bundle/run/collectd
+# Default path /usr/lib/splunk-otel-collector/agent-bundle on Linux.
+# Default path %ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle on Windows.
```
```suggestion
# Default path on Linux: "/usr/lib/splunk-otel-collector/agent-bundle".
# Default path on Windows: "%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle".
```

dmitryax commented on deployments/ansible/roles/collector/defaults/main.yml:
```diff
@@ -29,11 +31,18 @@ install_fluentd: true
+# Default path %ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle on Windows.
+agent_bundle_dir: ""
+
+# Default path /usr/lib/splunk-otel-collector/agent-bundle/run/collectd on Linux.
+# Default path %ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd on Windows.
```
```suggestion
# Default path on Linux: "/usr/lib/splunk-otel-collector/agent-bundle/run/collectd".
# Default path on Windows: "%ProgramFiles%\Splunk\OpenTelemetry Collector\agent-bundle\run\collectd".
```

dmitryax commented on deployments/ansible/roles/collector/tasks/vars.yml:
```diff
@@ -0,0 +1,65 @@
+      {%- if splunk_fluentd_config -%}
+        {{ splunk_fluentd_config }}
+      {%- else -%}
+        {{ansible_env.ProgramFiles}}/Splunk/OpenTelemetry Collector/fluentd/td-agent.conf
```
I see mix of `/` and `\` for Windows paths in the docs. Should we use `\` consistently in the docs and here?

dmitryax commented on deployments/ansible/roles/collector/tasks/win_fluentd_install.yml:
```diff
@@ -0,0 +1,23 @@
+- name: Install Fluentd on Windows
+  ansible.windows.win_package:
+    path: "{{fluentd_msi.dest}}"
+    state: present
```
Probably `notify: "restart windows fluentdwinsvc"` should be added as for otel collector

dmitryax reviewed (commented):
Looks good except for a few nits

Mukeshr-Ram commented:
Hello @dmitryax
Updated the PR with suggested changes.
Can you please review it.

dmitryax commented on deployments/ansible/README.md:
```diff
@@ -8,6 +8,25 @@ The role installs Splunk OpenTelemetry Connector configured to
+- Windows Server 2016 64-bit
+- Windows Server 2019 64-bit
+
+Ansible requires PowerShell 3.0 or newer and atleast .NET4.0 to be installed on Windows host.
```
```suggestion
Ansible requires PowerShell 3.0 or newer and at least .NET 4.0 to be installed on Windows host.
```

dmitryax commented:
@jcheng-splunk PTAL. Requested changes still blocks the merge

jcheng-splunk commented on deployments/ansible/README.md:
```diff
@@ -8,6 +8,25 @@ The role installs Splunk OpenTelemetry Connector configured to
+## Linux
+Currently, the following Linux distributions and versions are supported:
+
+- Amazon Linux
```
```suggestion
- Amazon Linux: 2
```

jcheng-splunk commented on deployments/ansible/README.md:
```diff
@@ -8,6 +8,25 @@ The role installs Splunk OpenTelemetry Connector configured to
+Currently, the following Linux distributions and versions are supported:
+
+- Amazon Linux
+- RedHat
```
```suggestion
- CentOS / Red Hat / Oracle: 7, 8
```

jcheng-splunk commented on deployments/ansible/README.md:
```diff
@@ -8,6 +8,25 @@ The role installs Splunk OpenTelemetry Connector configured to
+
+- Amazon Linux
+- RedHat
+- Debian: buster, stretch
```
```suggestion
- Debian: 9, 10
```

jcheng-splunk reviewed (commented):
LGTM after suggested doc updates.
